replica/table.cc

```diff
@@ -2089,9 +2096,41 @@ future<> tablet_storage_group_manager::update_effective_replication_map(const lo
     if (new_tablet_count > old_tablet_count) {
         tlogger.info0("Detected tablet split for table {}.{}, increasing from {} to {} tablets",
                         schema()->ks_name(), schema()->cf_name(), old_tablet_count, new_tablet_count);
-        return handle_tablet_split_completion(*old_tablet_map, *new_tablet_map);
+        co_await handle_tablet_split_completion(*old_tablet_map, *new_tablet_map);
+        co_return;
+    }
+
+    // Allocate/deallocate storage group if tablet is migrating in/out.
+    auto this_replica = locator::tablet_replica{
+        .host = erm.get_token_metadata().get_my_id(),
+        .shard = this_shard_id()
+    };
+    auto tablet_migrates_in = [this_replica] (locator::tablet_transition_info& transition_info) {
+        return transition_info.stage == locator::tablet_transition_stage::allow_write_both_read_old && transition_info.pending_replica == this_replica;
+    };
+    auto tablet_migrates_out = [&new_tablet_map, this_replica] (locator::tablet_id tid, locator::tablet_transition_info& transition_info) {
+        auto& current = new_tablet_map->get_tablet_info(tid).replicas;
+        auto is_this_replica = [this_replica] (const locator::tablet_replica& replica) { return replica == this_replica; };
+        return transition_info.stage == locator::tablet_transition_stage::end_migration &&
+            boost::find_if(current, is_this_replica) != current.end() &&
+            boost::find_if(transition_info.next, is_this_replica) == transition_info.next.end();
+    };
+    for (auto& transition : new_tablet_map->transitions()) {
+        auto tid = transition.first;
+        auto transition_info = transition.second;
+        if (!_storage_groups[tid.value()] && tablet_migrates_in(transition_info)) {
+            auto range = new_tablet_map->get_token_range(tid);
+            auto cg = std::make_unique<compaction_group>(_t, tid.value(), std::move(range));
+            _storage_groups[tid.value()] = std::make_unique<storage_group>(std::move(cg), &_compaction_groups);
+        } else if (_storage_groups[tid.value()] && tablet_migrates_out(tid, transition_info)) {
+            for (auto& cg_ptr : _storage_groups[tid.value()]->compaction_groups()) {
+                // Synchronizes with in-flight writes if any, and also takes care of flushing if needed.
+                co_await cg_ptr->stop();
+            }
+            _storage_groups[tid.value()] = nullptr;
```

Comment:
I"d exchange with nullptr before the stop loop above so not to leave the entry allocate with this storage group while it"s winding down